Speed up STEP file import

kpemartin opened the pull request:
This PR addresses Issue #18735 concerning slow import of STEP files.

Most of the slow performance is due to code which fetches the colour of each face in the imported objects.
This code makes up to three calls to `XPSDoc_ColorTool::GetColor` passing the `Topo_DS_Shape` for each such face. In the samples observed, the second call generally succeeds so usually two calls are made.

Internally `XPSDoc_ColorTool::GetColor` calls `XCAFDoc_ShapeTool::Search` to obtain a `TDS_Label` for the shape which is then used to find any colour information. It turns out that `XCAFDoc_ShapeTool::Search` is very slow, with execution time proportional to the number of (sub)shapes in the document (if I understand the code properly).

This PR tries to reduce this time by reducing the number of calls to `XCAFDoc_ShapeTool::Search`.

Ideally `XCAFDoc_ShapeTool::Search` should be fixed to not be so slow, but that is outside of FreeCAD's sphere of influence.

kpemartin committed:
Speed up colour fetching by mapping the shape to a label

XPSDoc_ColorTool::GetColor is much faster if given a TDS_Label instead of a Topo_DS_Shape because it does not have to call XCAFDoc_ShapeTool::Search (a very slow operation) each time.
By calling XCAFDoc_ShapeTool::Search to obtain the TDS_Label for each shape just once ourselves, XCAFDoc_ShapeTool::Search is only called once for each object instead of up to three times (for up to three calls to XPSDoc_ColorTool::GetColor)

kpemartin committed:
change creator methods to return doc object rather than stuffing it in a passed collection

kpemartin committed:
Roll tryPlacementFromMatrix into tryPlacementFromLoc

kpemartin committed:
Add a map from shape to Label (rough version) for fast colour lookup

maxwxyz commented:
moving to 1.2

maxwxyz commented:
@tritao wasn't boost removed?

kpemartin commented:
This is 3rd-party stuff, so even if Boost were removed from the main FC source, it might still be in third-party code.

In any case I suspect the problem is at my end. I think my file system took a hit a while ago and it looks like some of the LibPack files were damaged.

kpemartin committed:
Add adapter code to use unordered_map on TopoDS_Shape on older OCCT library

chennes commented on src/Mod/Import/App/ImportOCAF.cpp:
```diff
@@ -393,50 +409,44 @@ void ImportOCAF::createShape(
+                    ++n_faces;
+                    xp2.Next();
+                }
+                faceColors.resize(n_faces, color);
```
This ends up changing the default-fill logic, by filling the whole thing with *this face's color*, rather than the base color, doesn't it?

kpemartin replied:
Quick reply, as I'm about to leave on a 2-week trip. I recall considering the interaction of the overall colour and the face/edge/I forget the third thing colours. I'll have to take a closer look when I get back.

kpemartin replied:
You're correct, I'll fix this by using a different local variable for the per-face colour.

chennes replied:
Did you push this change? It looks like the problem persists?

kpemartin replied:
@chennes I have not pushed that change yet. I have just pushed it now after learning a bit more git-fu. I'm still cleaning up `Part::ShapeMapHasher`

chennes commented on src/Mod/Import/App/ImportOCAF.h:
```diff
@@ -54,6 +55,24 @@ namespace Part
 class Feature;
 }
 
+#if (OCC_VERSION_MAJOR < 7 || (OCC_VERSION_MAJOR == 7 && OCC_VERSION_MINOR < 8))
```
Can we use `Part::ShapeMapHasher` instead?

chennes reviewed (commented):
I took a short look at this today (code only, untested). A couple of questions below...

kpemartin commented:
There are three classes floating around for this file format: `ImportOCAF` seems to be original workhorse, `ImportOCAF2` seems to be a replacement, which is used when the user imports that file type and which uses `ImportOCAF` to do the guts of the work, and `ImportXCAF` which sort of seems to do the same thing on its own but does not appear to be called anywhere.

There is also an 'ImportOCAFCmd' class, also apparently not used

Nothing in FC itself references anything except `ImportOCAF2` (at least, doing a text search for the class names in the source and build folders) so I wonder what is the point in this somewhat messy structure? Are we trying to support outside code that links to FC code as a library and might use such classes?

chennes commented:
> Are we trying to support outside code that links to FC code as a library and might use such classes?

No. It's likely pure cruft related to the age of the codebase.

kpemartin committed:
Remove false asumption that different objects never hash the same
The old code was trying to record what shapes had been processed already by keeping a collection of the shapes' hashes. This would be incorrect if two different shapes happen to have the same hash, an unlikely but possible event.
This also removes a reference to Part::ShapeMapHasher

kpemartin committed:
Remove unused classes ImportOCAFCmd and ImportXCAF

kpemartin committed:
Delete unreferenced/unbuilt ImportOCAFAssembly files

kpemartin committed:
Remove ineffective color-load code from Part/ImportStep
This code has been ineffective since May 14 2016 commit bca7a1866e596ad2e318809f35b6d08776226a69 when OCC 7.0 support was added and the colour-loading code was deleted (presumably to be re-impleted later). But it appears thet Part/ImportStep is deprecated anyway.
This change is just pruning out more of the scar tissue to remove a reference to Part::ShapeMapHasher

kpemartin committed:
Remove use of Part::ShapeMapHasher and improve performance
Using TopExp::MapShapesAndAncestors can be an expensive way to
find ancestors of a particular object.
First it builds a map from all shapes in containingShape regardless
of type, to all the ancestors of (each of those shapes which are of
desiredAncestorType),
Then we pick out the single ancestor list for descendentShape,
and have to remove duplicates.
OCC does not give a better way of finding all the ancestors for a
specific shape.
This can be expensive if containingShape is particularly complex.
Instead we use code which is essentially the guts of
TopExp::MapShapesAndAncestors pruned down to find the ancestors
for a single shape. This code has the same double-nested loop
but it has a quick exit once we've found an ancestry connection
for a particular ancestor, and it does not waste time building the
collections of ancestore for other shapes.

kpemartin committed:
Use a set of non-free edges rather than their hashes
(also removes last use of Part::ShapeMapHasher)

kpemartin committed:
Remove ShapeMapHasher.h file
Any (remaining) users of ShapeMapHasher now use std::hash<TopoDS_SHape>

kpemartin commented:
This latest push is almost but not quite done. It may fail Ubuntu build again and I have to find a better place to put the specialization of `std::hash<TopoDS_Shape>` to resolve this.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

kpemartin committed:
Move std::hash<TopoDS_Shape> specialization to central location
It is needed for Ubuntu builds by both the Import and Part modules

kpemartin committed:
Remove overlapping memcpy with undefined results
The old code was making a view on a subequence of a std::string and
assigning that back onto the same string, causing an overlapping memcpy
(noticed looking at warnings of CI build)

chennes commented:
@kpemartin is this ready for review now? The ubuntu CI passed, at any rate.

kpemartin commented:
@chennes Apparently not, I just tried opening a step file and ended up with an empty drawing! I guess some debugging is in order...

kpemartin commented:
Having removed `ImportOCAFCmd` and `ImportXCAF` I looked into what all the rest of `ImportOCAF2` was for, and it is a completely different importer which refers to `ImportOCAF` (the class I've been optimizing) as the "legacy" importer. As it turns out, if you turn on the innocuously-named "Use LinkGroup" import option, this actually means to use the entirely different import code in `ImportOCAF2`

The latter code seems to do (almost) what the legacy code does and a lot more, implementing all the importing options. It also has effectively the same colour-setting speedup I added to ImportOCAF. The only thing not implemented is the don't-use-LinkGroup choice; instead of just doing that it punts to the legacy importer.

Now I feel I've been wasting my time optimizing the legacy importer, and I should have just modified the new importer to not use LinkGroups for Assemblies but instead `App::Part` like the legacy importer does (and also use `App::Part` rather than `Part::Compound2` for decomposed compound objects). There could, however, be some more subtle regressions involved.

kpemartin commented:
...aaaand it looks like autocomplete gave me a wrapper for a method (`resize_and_overwrite`) of `std::string` not available until C++23. The Windows environment provides this under the name `_Resize_and_overwrite` and I didn't notice autocomplete giving me this.

kpemartin commented:
I had a look at modifying the "new" importer `ImportOCAF2` to reproduce the results of old importer (`importOCAF`) when the "Use LinkGroup" import option is not set rather than just punting to the old importer.
It didn't go well.
Instead I'll keep these speedups in the old importer and the hash/collection support upgrades, but see about changing the preferences to discourage use of the old importer (with the goal of eliminating it entirely in the distant future). This would involve restating the "Use LinkGroup" import option as a "Use legacy importer" with the boolean value reversed, and changing the default preferences to not use the legacy importer. I can also see if it is possible to mark some of the import options as being unimplemented and disabling them when the legacy importer is used (I believe this was done for the DXF importers).

kpemartin committed:
Rebase including OCCT upgrade

chennes commented:
I rebased and fixed the merge conflict. I also dropped the doubling of the contents of the Py::List due to constructing it with an integer parameter and then inserting those elements again. You had:
```c++
Py::List list(uniqueIndices.size());
for (int idx : uniqueIndices) {
    list.append(...);
}
```
That initial constructor doesn't just reserve the space, it actually creates the elements, so the `append()` was adding them *after* a bunch of `None`s.

Unfixed, but we should think about it: this PR changes `ancestorsOfType`'s result ordering. Main iterates `findAncestors` output and de-duped as it went, preserving order; this PR iterates a `std::unordered_set<int>`, so the returned order is now unspecified. Is that intentional/desirable?

kpemartin commented:
@chennes I'm not quite sure if your concern is that the order is officially undefined, or that it may be different from what it was x months ago in a previous version of FC.

The order returned by `findAncestors` in Head is not well-defined (that is, no documentation says what order it returns), but it at least turns out to be consistent if called multiple times on the same object (not because the underlying code that lists the ancestors is itself well-defined but because this result is cached). I believe there is, however, no assurance that you will get the same order when called on a copy of the object, or across a save/close document/load document cycle, or if the object is modified in a manner that should not affect the contents of the ancestors list for a particular sub-object.

I don't see that the undocumented order returned by the `unordered_set` version makes any difference as opposed to the undocumented order order returned by `findAncestors`. The only way it would matter would be if some code got the ancestors both through the Python method and also by calling the C++ method directly, expecting the two to be in the same order.

The point of this particular change was to avoid building the result list piecemeal, so it could all be allocated at once in its final size. As you observe, this code was incorrect, and should perhaps have used its own increasing index to index into the result list to place the elements therein, but that introduces another iteration variable. It is not a big win, performance-wise, so if the different ordering is truly a concern and there is no way to do a pre-sized return list without code clumsiness, that could be skipped entirely.

In any case, my rebase attempt from June left me unable to build because of all the library reference changes included in Head and (it seemed at the time) inconsistencies between the libPack and what the code needed—I don't recall the details—all of which took the wind from my sails. Over the next few days I'll see if I can get the build environment working and review my PR again. The change for pre-sizing the result list is not critical and I'll just remove it if it makes the code too clumsy, or if you want the same order as `findAncestors` returns—whatever that is.

chennes commented:
> The point of this particular change was to avoid building the result list piecemeal, so it could all be allocated at once in its final size. As you observe, this code was incorrect, and should perhaps have used its own increasing index to index into the result list to place the elements therein, but that introduces another iteration variable.

If you want to achieve this with a minimum code change, just use the `[]` accessor to fill the list instead of `append()`, and restore the original size statement. You just can't do both, you double the size of the list and leave the front half full of `None`s.

Regarding the ordering: my concern is just that it used to be deterministic, and no longer is. If no one cares then it's fine with me.

kpemartin commented:
There are actually hundreds of occurrences of the pattern
```
Py::List x;
for (something) {
    x.append(interationValue);
}
```
and the code in `TopoShapePy::ancestorsOfType` is just one of them.

I have just created #32091 to address this generally, and will not do anything about it in this PR.
